src/rand.rs

```diff
@@ -0,0 +1,139 @@
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+//! Random number generator based of the PCG paper (http://www.pcg-random.org/paper.html).
+
+use std::u32;
+use std::time::*;
+
+pub struct Rng {
+    state: u64,
+    inc: u64,
+}
+
+impl Default for Rng {
+    fn default() -> Self {
+        match SystemTime::now().duration_since(UNIX_EPOCH) {
+            Ok(res) => Self::seed_with(res.as_secs()),
+            Err(_) => Self::seed_with(6_364_136_223_846_793_005)
+        }
+    }
+}
+
+impl Rng {
+    fn pcg32(&mut self) -> u32 {
+        let oldstate = self.state;
+        // Advance internal state
+        self.state = u64::wrapping_add(u64::wrapping_mul(oldstate, 6_364_136_223_846_793_005u64), self.inc | 1);
+        // Calculate output function (XSH RR), uses old state for max ILP
+        let xorshifted = (((oldstate >> 18) ^ oldstate) >> 27) & 0xFFFF_FFFF;
+        let rot = (oldstate >> 59) & 0xFFFF_FFFF;
+        let v = (xorshifted >> rot) | (xorshifted << (u64::wrapping_sub(0, rot) & 31));
+        v as u32
+    }
+
+    pub fn seed_with(seed: u64) -> Self {
+        // We xor the seed with a randomly chosen number to avoid ending up with
+        // a 0 state which would be bad.
+        Self {
+            state: seed ^ 0xedef_335f_00e1_70b3,
+            inc: 12345,
+        }
+    }
+
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    #[allow(should_implement_trait)]
+    pub fn next(&mut self) -> u32 {
+        self.pcg32()
+    }
+
+    pub fn next_in_range(&mut self, min: u32, max: u32) -> u32 {
+        (self.pcg32() % (max - min)) + min
+    }
+
+    pub fn next_double(&mut self) -> f64 {
+        let max = f64::from(u32::MAX);
+        let n = f64::from(self.next());
+        n / max
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::collections::BTreeMap;
+
+    use super::Rng;
+
+    #[test]
+    fn avg_median() {
+        let mut rng = Rng::new();
+        let mut numbers = vec![];
+        for _ in 0..1_000_000 {
+            numbers.push(rng.next());
+        }
+        let (min, max) = {
+            (numbers.iter().map(|&num| num).min().expect("minimum"), numbers.iter().map(|&num| num).max().expect("maximum"))
+        };
+        let median = ((max - min) / 2) as f64;
+        let len = numbers.len();
+        let avg = numbers.into_iter().map(|num| num as u64).sum::<u64>() / len as u64;
+        let ratio = median / avg as f64;
+        assert!(ratio - 1.0 < 0.01);
```

Comment:
There seems to be potential for a false negative when `(median / avg) < 1.0` and `ratio - 1.0 < -0.01`. It could be fixed by using the absolute value of `ratio - 1.0`